Concurrent table scans

sdd opened the pull request:
This is a bit of an experiment to see how things could look if we tried to:

- process Manifest Lists and Manifest Files concurrently rather than sequentially
- process the stream of file scan tasks concurrently, streaming record batches from multiple files at the same time.

I'd like to add some unit tests to confirm that this behaves as expected beyond the existing tests that we have for TableScan, and add an integration / performance test that can quantify any performance improvements (or regressions 😅 ) that we get from these changes.

Let me know what you all think.

marvinlanhenke commented on crates/iceberg/src/scan.rs:
```diff
@@ -189,66 +195,20 @@ impl TableScan {
-                    partition_spec_id,
-                    &context,
-                )?;
+        spawn(async move {
```
do we need to `spawn` here or is the `try_for_each_concurrent` in `run(...)` already enough?

sdd replied:
I went with using an mpsc channel to avoid having 2 nested try_for_each_concurrent macros. On reflection the inner try_for_each_concurrent is probably unnecessary, since there's only one async operation needed for each manifest file, despite that manifest file producing n DataFileTasks. So I can probably ditch the channel too. The channel in reader.rs is also overkill as I've only got a single try_for_each_concurrent there.

I just wanted to get the ball rolling on this while we're waiting for the filtering code to get signed off, it's fun writing this kind of stuff 😁

sdd replied:
I reverted reader.rs to be essentially the same as before, without any concurrent processing of batches from within the same file. I removed the nested try_for_each_concurrent from scan.rs but kept the mpsc channel.

marvinlanhenke commented on crates/iceberg/src/scan.rs:
```diff
@@ -302,13 +262,147 @@ impl TableScan {
+}
+
+#[derive(Debug)]
+struct ConcurrentFileScanStreamContext {
```
I think once we have a runtime and the async approach is approved, we can get rid of the `FileScanStreamContext` and merge the struct into `ConcurrentFileScanStreamContext`

sdd replied:
Yeah, completely agree. Kept them separate for now as it made it easier to have the old non-concurrent version in the code base at the same time.

sdd replied:
I've refactored these into a single context. Now that we are using the `ExpressionEvaluator` though, I ended up creating another context object that gets passed to the manifest file closure. This was just a tuple before but clippy complained when I added one more item to it :-)

marvinlanhenke commented on crates/iceberg/src/scan.rs:
```diff
@@ -302,13 +262,147 @@ impl TableScan {
+                CONCURRENCY_LIMIT_MANIFEST_FILES,
+                Self::process_manifest_file,
+            )
+            .await
```
will this yield the first FileScanTask when its available - or do we have to iterate over the complete stream, before we can return any result?

sdd replied:
Yes, everything proceeds fully concurrently - whatever manifest is ready first will start yielding tasks up to the stream first

marvinlanhenke replied:
thanks for explaining

marvinlanhenke commented on crates/iceberg/src/scan.rs:
```diff
@@ -302,13 +262,147 @@ impl TableScan {
+                CONCURRENCY_LIMIT_MANIFEST_ENTRIES,
+                Self::process_manifest_entry,
+            )
+            .await
```
we had a discussion on this in #124 (in case you missed it) to not go overboard with the task spawning

sdd replied:
Task spawning concurrency is limited by the `CONCURRENCY_LIMIT_MANIFEST_ENTRIES` constant, which I'm happy to adjust in this PR to suit

marvinlanhenke reviewed (commented):
Thanks @sdd for doing this draft.

Just in case you missed it we had a similiar discussion in #124 about a possible approach.
Left some comments (mostly questions). 

In general I think we should wait for the runtime and the Evaluators (Manifest, Expression, etc.) to land. Add more tests in scan.rs - and then refactor into the async/ multi-threaded version. Curious for more comments on this, though.

sdd commented:
I've updated this to ditch the concurrency when processing `ManifestEntry` items within a single `Manifest`, producing them asynchronously but sequentially instead. I've kept the limited concurrency when processing `ManifestFile`s within the scan's snapshot's `ManifestList`.

I've kept the approach of using an mpsc channel with a spawned task, with that task using `try_for_each_concurrent` to achieve the concurrency. This is because without the channel and spawned task, we'd need to use an async closure, which is unstable rust. With the spawned task we only need to use an async block, which is in stable rust.

sdd commented on crates/iceberg/src/scan.rs:
```diff
@@ -389,12 +333,158 @@ impl FileScanStreamContext {
+        });
+
+        futures::stream::iter(filtered_manifest_files)
+            .try_for_each_concurrent(
```
Aah. You're right. `try_for_each_concurrent` is concurrent but not parallel. What I have should still be an improvement on what we have already, as we'll at least be able to have multiple fileIO file requests in-flight at the same time, but this is not fully parallelised.

I'll work on making this parallel as well as concurrent. Some initial experimentation shows that it is tricky to please the borrow checker, but I'll keep trying.

sdd replied:
Thanks. I'm not happy with it now that you pointed out the issue with `try_for_each_concurrent`. I'm in the middle of refactoring it and should have something quite soon.

sdd replied:
Hey @liurenjie1024 - I think I'm happy with this now. Let me know what you think.

There is one issue remaining. It seems that our `crate::runtime::spawn` does not behave exactly the same as `tokio::spawn` for tasks that are passed an async function that return a `Result`? I could not get this to work with `crate::runtime::spawn` but it does with `tokio::spawn`. @odysa could you help?

odysa commented:
@sdd `JoinHandle` in `Tokio` and `async-std` have different return types.
In [Tokio](https://docs.rs/tokio/latest/src/tokio/runtime/task/join.rs.html#324)
```rs
impl<T> Future for JoinHandle<T> {
    type Output = super::Result<T>;
```
In [Async-std](https://github.com/async-rs/async-std/blob/b7be2183d55a6bd62907cac0b10732f377d39fe2/src/task/join_handle.rs#L79)
```rs
impl<T> Future for JoinHandle<T> {
    type Output = T;
```
`crate::runtime::spawn` is consistent with async-std, so you may remove the `?` to resolve this issue.

However, we can still make it to behave like Tokio apis. We may need to discuss it further. cc @liurenjie1024 @Xuanwo

sdd commented:
Thanks @odysa - I must be going crazy, I thought I tried that already but you were right, that worked! 👍🏼

Xuanwo commented on crates/iceberg/src/scan.rs:
```diff
@@ -197,134 +193,103 @@ impl<'a> TableScanBuilder<'a> {
+                CONCURRENCY_LIMIT_MANIFEST_ENTRIES,
+                |(manifest_entry_context, tx)| async move {
+                    crate::runtime::spawn(async move {
+                        Self::process_manifest_entry(manifest_entry_context, tx).await
```
As a user, I prefer to manage logic like this myself because I have my own pipeline with specific resource allocation and concurrency limits. So I think @liurenjie1024's concern is true. Maybe we can make those behavior optional and disable by default. Users who want iceberg-rust to do this can enable with some config or flags.

sdd replied:
If I make the concurrncy limits configurable in the builder as @liurenjie1024 suggested above, then we could have a `with_no_concurrency` method on the builder. Would this be acceptable?

sdd replied:
In terms of having non-concurrent as the default option, my preference would be for the library to be able to choose the optimally performant (probably the number of cores?) as the default rather than no concurrncy. No concurrncy feels like a more advanced, non-default choice. @liurenjie1024 what are your thoughts?

sdd replied:
Regarding the large number of tasks - there are a large number generated over the duration of the plan, but only a fixed small number that are in existence at one time.

sdd replied:
Yeah you're right, `with_no_concurrency()` is a bit pointless when you can just set `with_max_concurrency(1)`! 😂

Xuanwo commented on crates/iceberg/src/expr/predicate.rs:
```diff
@@ -668,6 +668,10 @@ pub enum BoundPredicate {
 
+/// Newtype to prevent accidentally confusing predicates that are bound to a partition with ones that are bound to a schema.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct PartitionBoundPredicate(pub(crate) BoundPredicate);
```
Yes, I don't think the confusion is significant enough to require a new-type.

sdd replied:
I've removed it. It was ugly anyway 😅

sdd commented on crates/iceberg/src/scan.rs:
```diff
@@ -197,134 +193,103 @@ impl<'a> TableScanBuilder<'a> {
-                }
+        // used to stream ManifestEntryContexts between stages of the file plan operation
+        let (manifest_entry_ctx_tx, manifest_entry_ctx_rx) =
+            channel(CONCURRENCY_LIMIT_MANIFEST_FILES);
```
Absolutely, I'd considered doing that, I think it's a good idea. Will add

sdd replied:
Done - let me know if you're happy with the approach

sdd commented on crates/iceberg/src/spec/manifest_list.rs:
```diff
@@ -79,6 +79,11 @@ impl ManifestList {
     }
+
+    /// Take ownership of the entries in the manifest list, consuming it
+    pub fn consume_entries(self) -> Box<dyn Iterator<Item = ManifestFile>> {
```
Done, thanks!

sdd commented on crates/iceberg/src/scan.rs:
```diff
@@ -338,157 +303,463 @@ impl TableScan {
+            let Some(ref partition_bound_predicate) =
+                manifest_entry_context.partition_bound_predicate
+            else {
+                panic!("partition_bound_predicate expected to always be present if snapshot_bound_predicate is present");
```
Good suggestion - done!

sdd commented:
PTAL @liurenjie1024 and @Xuanwo: I've addressed your helpful feedback and I think this is now ready. Performance tests of this on my performance testing branch show clear gains when scanning snapshots whose manifest list contains multiple manifests. I'm really happy with this now!

ZENOTME commented:
> PTAL @liurenjie1024 and @Xuanwo: I've addressed your helpful feedback and I think this is now ready. Performance tests of this on my performance testing branch show clear gains when scanning snapshots whose manifest list contains multiple manifests. I'm really happy with this now!

I'm curious about the performance gain from this. Can you share the related data?

sdd commented:
Hey @ZENOTME! Sure. If you check out my `perf-suite` branch from [my other PR](https://github.com/apache/iceberg-rust/pull/497), this is branched off main and can be used to get a baseline of `plan_files` performance without the changes in this PR. You'll need to run `cargo install just` if you don't have `just` already installed, and then `just perf-run`. This should create the docker setup for the performance testing environment, download some test data fro the NYC Taxi dataset, create an Iceberg table inside the perf testing env, insert the test data into the table, and then run the perf tests.

The perf tests themselves consist of four scenarios that each execute `plan_files` in different scenarios:
* a file plan for a scan that reads **all** rows from **all** data files in the current snapshot (ie no filter);
* a file plan for a scan that reads **some** rows from **all** data files in the current snapshot (`passenger_count` = 1);
* a file plan for a scan that reads **all** rows from **one** data file in the current snapshot (filter: `tpep_pickup_datetime` = '2024-02-01')
* a file plan for a scan that reads **some** rows from **one** data file in the current snapshot (filter: `tpep_pickup_datetime` = '2024-02-01' AND `passenger_count` = 1);

The benchmarks get ran using criterion. Here's an example of the test output when I ran it just now on my M1 Pro Macbook:

<img width="681" alt="image" src="https://github.com/user-attachments/assets/3d521b80-b3ab-4421-8889-e39cc1a0ec91">

Now, if you cherry-pick the commit from this concurrent table scan branch into the `perf-suite` branch:

```bash
git cherry-pick 8eef484094dbb7c55ac3181bbd552090308591c9
```

And re-run the performance tests, you'll see something similar to the following:

<img width="689" alt="image" src="https://github.com/user-attachments/assets/28e41e05-b1e2-48a9-a26c-20ca940e688a">

As you can see the times come down from around 1.5s to around **0.5s**, a big improvement!

In fact, if you then cherry-pick the commit from my [follow-on PR with the object cache](https://github.com/apache/iceberg-rust/pull/512):

```bash
git cherry-pick c696a3f9a7dacf822e6d5bc76f13d24e0b50ee31
```


and re-run the performance tests again, you'll see something even more remarkable:

<img width="578" alt="image" src="https://github.com/user-attachments/assets/12e91179-2ea3-4dad-9581-d7698b80b653">

Now that the retrieval and parsing of the `Manifest` and `ManifestList`s are taking place, the first run in the performance test will take the same time as above but all of the subsequent runs that are averaged to get an average time for each test are much, much faster, reducing the average time from 500ms to about 1.5ms. 🤯

Xuanwo commented:
> Now that caching of the retrieval and parsing of the `Manifest` and `ManifestList`s are taking place, the first run in the performance test will take the same time as above but all of the subsequent runs that are averaged to get an average time for each test are much, much faster, reducing the average time from 500ms to about 1.5ms. 🤯

Great!

ZENOTME commented:
> Hey @ZENOTME! Sure. If you check out my `perf-suite` branch from [my other PR](https://github.com/apache/iceberg-rust/pull/497), this is branched off main and can be used to get a baseline of `plan_files` performance without the changes in this PR. You'll need to run `cargo install just` if you don't have `just` already installed, and then `just perf-run`. This should create the docker setup for the performance testing environment, download some test data fro the NYC Taxi dataset, create an Iceberg table inside the perf testing env, insert the test data into the table, and then run the perf tests.
> 
> The perf tests themselves consist of four scenarios that each execute `plan_files` in different scenarios:
> 
> * a file plan for a scan that reads **all** rows from **all** data files in the current snapshot (ie no filter);
> * a file plan for a scan that reads **some** rows from **all** data files in the current snapshot (`passenger_count` = 1);
> * a file plan for a scan that reads **all** rows from **one** data file in the current snapshot (filter: `tpep_pickup_datetime` = '2024-02-01')
> * a file plan for a scan that reads **some** rows from **one** data file in the current snapshot (filter: `tpep_pickup_datetime` = '2024-02-01' AND `passenger_count` = 1);
> 
> The benchmarks get ran using criterion. Here's an example of the test output when I ran it just now on my M1 Pro Macbook:
> 
> <img alt="image" width="681" src="https://private-user-images.githubusercontent.com/323439/354498323-3d521b80-b3ab-4421-8889-e39cc1a0ec91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI3NTIyOTQsIm5iZiI6MTcyMjc1MTk5NCwicGF0aCI6Ii8zMjM0MzkvMzU0NDk4MzIzLTNkNTIxYjgwLWIzYWItNDQyMS04ODg5LWUzOWNjMWEwZWM5MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwNFQwNjEzMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMzY3NDhkNTFmMjQ3ZDQyNWE0Yjk0NzY3NzVlNjJjODY4NTcxZTY3MTg4NmNlMjA3MGY1ZDRkMzlmNjJiMDk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bi41R1qt3HLdJjxiQPDf5U1F3WMhEdby2bEpZDTdIgw">
> Now, if you cherry-pick the commit from this concurrent table scan branch into the `perf-suite` branch:
> 
> ```shell
> git cherry-pick 8eef484094dbb7c55ac3181bbd552090308591c9
> ```
> 
> And re-run the performance tests, you'll see something similar to the following:
> 
> <img alt="image" width="689" src="https://private-user-images.githubusercontent.com/323439/354499847-28e41e05-b1e2-48a9-a26c-20ca940e688a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI3NTIyOTQsIm5iZiI6MTcyMjc1MTk5NCwicGF0aCI6Ii8zMjM0MzkvMzU0NDk5ODQ3LTI4ZTQxZTA1LWIxZTItNDhhOS1hMjZjLTIwY2E5NDBlNjg4YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwNFQwNjEzMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MWYzOTQxYmMyMDEyZjk5NWUwZTE2NzFjODRkYTkwYzA5MzAyMzM1NjlhOWZlYzgxOGIxODA0Y2I5YTE3YWYwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.WLrE2Rybmfyg4Qn3GmIeIuDmXLfESuXpM-UTvQ26LH4">
> As you can see the times come down from around 1.5s to around **0.5s**, a big improvement!
> 
> In fact, if you then cherry-pick the commit from my [follow-on PR with the object cache](https://github.com/apache/iceberg-rust/pull/512):
> 
> ```shell
> git cherry-pick c696a3f9a7dacf822e6d5bc76f13d24e0b50ee31
> ```
> 
> and re-run the performance tests again, you'll see something even more remarkable:
> 
> <img alt="image" width="578" src="https://private-user-images.githubusercontent.com/323439/354501480-12e91179-2ea3-4dad-9581-d7698b80b653.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI3NTIyOTQsIm5iZiI6MTcyMjc1MTk5NCwicGF0aCI6Ii8zMjM0MzkvMzU0NTAxNDgwLTEyZTkxMTc5LTJlYTMtNGRhZC05NTgxLWQ3Njk4YjgwYjY1My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwNFQwNjEzMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNmI0MDg4OGI4YWQ3N2FiNjFjNTljODgzMzI2MjA2MWJlYjBhYzE3MmI0M2Q1NmMzNWI1MjczZmFlN2E3MTM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.WrzRg34n0axA3cur2dUnbvSf1tWBD81aYRaI3HKgCZE">
> Now that caching of the retrieval and parsing of the `Manifest` and `ManifestList`s are taking place, the first run in the performance test will take the same time as above but all of the subsequent runs that are averaged to get an average time for each test are much, much faster, reducing the average time from 500ms to about 1.5ms. 🤯

Thanks @sdd! That's awesome! BTW, can we integrate the performance test within the repository so that it will be helpful for performance improvement in the future? cc @liurenjie1024 @Xuanwo @sdd

sdd commented:
Sorry guys, been off-grid for a few days. I've addressed all of your suggestions @liurenjie1024, thanks.

> > Thanks @sdd! That's awesome! BTW, can we integrate the performance test within the repository so that it will be helpful for performance improvement in the future? cc @liurenjie1024 @Xuanwo @sdd
> 
> I agree that a benchmark would be helpful for validating this, but I would prefer to do this after integrating with a sql engine like datafusion, which makes maintaince easier.

@liurenjie1024  Why don't we just bring in the performance tests as they are / with whatever changes are needed from review, and we can always delete them and add different ones once we have the SQL engine in place? It's not much effort to delete them when they've been superseded but it is very useful to have them in place now whilst a lot of performance-related changes are being done. I'm happy to get the perf tests in my other PR into shape to merge as @ZENOTME suggests, even if they will be replaced at some point in the future.

sdd commented:
@liurenjie1024 @Xuanwo are we able to merge this now?

SteveLauC commented on crates/iceberg/src/scan.rs:
```diff
@@ -55,17 +55,23 @@ pub struct TableScanBuilder<'a> {
 
 impl<'a> TableScanBuilder<'a> {
     pub(crate) fn new(table: &'a Table) -> Self {
+        let num_cpus = num_cpus::get();
```
Not a reivew comment, just a glimpse of this PR, we can replace this with `std::thread::available_parallelism()`, `num_cpus` is almost unmaintained (as it no longer needs to), and has issues that are solved by the std counterpart, e.g., this one: [std::thread::available_parallelism() return 0](https://github.com/rust-lang/rust/issues/115868).

xxchan commented on crates/iceberg/src/scan.rs:
```diff
@@ -199,134 +219,119 @@ impl<'a> TableScanBuilder<'a> {
+            snapshot_schema: schema,
             case_sensitive: self.case_sensitive,
-            filter: self.filter.map(Arc::new),
+            predicate: self.filter.map(Arc::new),
+            snapshot_bound_predicate: snapshot_bound_predicate.map(Arc::new),
```
May I ask what does the "snapshot" mean here in `snapshot_bound_predicate`? :eyes:
